### DISCUSSION TITLE
Postgres error returned from database: column \"id\" does not exist. Table with mixed name cases.

### DISCUSSION THREAD

**Discussion by gitNemo** [rbatis/rbatis#291] [General]
### **Versions**

- **Rust**: 1.50
- **rbatis**: 1.8
- **Database**: Postgres 9.6
- **Operating System**: macOS 10.15.7

### **Problem Description**

I am working with a legacy production database for an existing product within which there are a number of tables and columns with mixed names cases like "COM_Item" as a table name and "Oid" as field name.
So my struct defined in Rust is like this;
```rust
#[crud_enable(id_name:"oid"|table_name:"COM_Item")]
#[derive(Clone, Debug)]
pub struct ComItem {
    pub oid: Option<i8>,
    pub label1: Option<String>,
}
```

- The Database complains with: **error returned from database: relation \\"com_item\\" does not exist** or **column \"oid\" does not exist**
_Which I have to convert the table name in lower case, but the database is not mine and there is others ecosystems depending on it. Not an elegant way to do it._

How to specify tables names as in Database, without converting them to lower case ?

**Comment by zhuxiujia** [comment]
> ### **Setup**
> ### **Versions**
> * **Rust**: 1.50
> * **rbatis**: 1.8
> * **Database**: Postgres 9.6
> * **Operating System**: macOS 10.15.7
> 
> ### **Problem Description**
> I am working with a legacy production database for an existing product within which there are a number of tables and columns with mixed names cases like "COM_Item" as a table name and "Oid" as field name.
> So my struct defined in Rust is like this;
> 
> ```rust
> #[crud_enable(id_name:"oid"|table_name:"COM_Item")]
> #[derive(Clone, Debug)]
> pub struct ComItem {
>     pub oid: Option<i8>,
>     pub label1: Option<String>,
> }
> ```
> 
> * The Database complains with: **error returned from database: relation \"com_item\" does not exist** or **column "oid" does not exist**
>   _Which I have to convert the table name in lower case, but the database is not mine and there is others ecosystems depending on it. Not an elegant way to do it._
> 
> How to specify tables names as in Database, without converting them to lower case ?




i  see  your define  table_name:"COM_Item"
```rust
#[crud_enable(id_name:"oid"|table_name:"COM_Item")]
#[derive(Clone, Debug)]
pub struct ComItem {
    pub oid: Option<i8>,
    pub label1: Option<String>,
}
```
this define COM_Item will  be  Used in SQL。
This code does seem to work，Did you use Cargo Update to update to the latest version of ``` rbatis ```?

**Comment by zhuxiujia** [comment]
you can use log library to see rbatis makes sql data

use`  fast_log::init_log("requests.log", 1000, log::Level::Info, None, true); `
to show rbatis sql log

**Comment by gitNemo** [comment]
@zhuxiujia
Yep!, the table name is COM_Item
Did the update to 1.8.77 in a minute.
And here is the log with fast_log after the update;
```
[rbatis] [] Query ==> select oid,label1 from com_item 
[rbatis] [] Args  ==> []
[rbatis] [] ReturnErr  <== error returned from database: relation "com_item" does not exist ...
```

**Comment by zhuxiujia** [comment]
@gitNemo 

```rust
    #[crud_enable(id_name:"oid"|table_name:"COM_Item")]
    #[derive(Clone, Debug)]
    pub struct ComItem {
        pub oid: Option<i8>,
        pub label1: Option<String>,
    }

    #[async_std::test]
    pub async fn test_save() {
        fast_log::init_log("requests.log", 1000, log::Level::Info, None, true);
        let rb = Rbatis::new();
        rb.link("mysql://root:123456@localhost:3306/test")
            .await
            .unwrap();
        let activity = ComItem {
            oid: Some(1),
            label1: Some("1".to_string())
        };
        let r = rb.save("", &activity).await;
        if r.is_err() {
            println!("{}", r.err().unwrap().to_string());
        }
    }

```
* result
```rust
2021-03-07 01:20:11.777386700 +08:00 INFO rbatis::plugin::log - [rbatis] [] Exec  ==> insert into COM_Item (oid,label1) values (?,?)
                                                                [rbatis] [] Args  ==> [1,"1"]
```

**Comment by gitNemo** [comment]
@zhuxiujia 
Mea culpa! 
table_name was "com_item", changed it to "COM_Item" but still has those errors;
Changed fields names in the struct to TitleCase because fields are in TitleCase too, for testing;

```
[rbatis] [] Exec  ==> insert into COM_Item (Oid,Label1) values ($1,$2)
[rbatis] [] Args  ==> [1,"DDDD"]
[rbatis] [] ReturnErr  <== error returned from database: relation "com_item" does not exist...

[rbatis] [] Query ==> select Oid,Label1 from COM_Item 
[rbatis] [] Args  ==> []
[rbatis] [] ReturnErr  <== error returned from database: relation "com_item" does not exist...  
```

Just a question;
Do I have to specify all the fields in the struct with all the relations? there is 120 in total !

**Comment by zhuxiujia** [comment]
You don't need to specify them all if some columns are not needed

**Comment by gitNemo** [comment]
@zhuxiujia 
This is weird, when I change the table name from the database to **com_item** all lower case, I get this error;
**relation "COM_Item" does not exist.**

And by returning it to **COM_Item**, I get this error;
**relation "com_item" does not exist**

Am using a Postgres database version 9.6.

**Comment by zhuxiujia** [comment]
2.SQL statement is not sensitive to case, uppercase and lowercase SQL statement meaning is the same, all uppercase converted to lowercase for interpretation,

Select * from NODE and select * from NODE are identical;



3. To query or call uppercase PostgreSQL objects, enclose the corresponding name in double quotation marks, such as "NODE";



4. One final tip is to use lowercase in PostgreSQL, since "NODE" feels weird in double quotation marks.

> @zhuxiujia
> This is weird, when I change the table name from the database to **com_item** all lower case, I get this error;
> **relation "COM_Item" does not exist.**
> 
> And by returning it to **COM_Item**, I get this error;
> **relation "com_item" does not exist**
> 
> Am using a Postgres database version 9.6.

1.PostgreSQL is case-sensitive to the names of established objects (stored procedures, functions, tables, fields, sequences, etc.);

2.SQL statement is not sensitive to case, uppercase and lowercase SQL statement meaning is the same, all uppercase converted to lowercase for interpretation,

Select * from NODE and select * from NODE are identical;

3. To query or call uppercase PostgreSQL objects, enclose the corresponding name in double quotation marks, such as "NODE";

4. One final tip is to use lowercase in PostgreSQL, since "NODE" feels weird in double quotation marks.
